replicator/changes.go

```diff
@@ -0,0 +1,171 @@
+package replicator
+
+import (
+	"context"
+	"time"
+
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/db"
+)
+
+// Used by BLIP connections for changes.  Supports both one-shot and continuous changes.
+func generateBlipSyncChanges(database *db.Database, inChannels base.Set, options db.ChangesOptions, docIDFilter []string, send func([]*db.ChangeEntry) error) (err error, forceClose bool) {
+
+	// Store one-shot here to protect
+	isOneShot := !options.Continuous
+	err, forceClose = GenerateChanges(context.Background(), database, inChannels, options, docIDFilter, send)
+
+	// For one-shot changes, invoke the callback w/ nil to trigger the 'caught up' changes message.  (For continuous changes, this
+	// is done by MultiChangesFeed prior to going into Wait mode)
+	if isOneShot {
+		_ = send(nil)
+	}
+	return err, forceClose
+}
+
+// Shell of the continuous changes feed -- calls out to a `send` function to deliver the change.
+// This is called from BLIP connections as well as HTTP handlers, which is why this is not a
+// method on `handler`.
+func GenerateChanges(cancelCtx context.Context, database *db.Database, inChannels base.Set, options db.ChangesOptions, docIDFilter []string, send func([]*db.ChangeEntry) error) (err error, forceClose bool) {
```

Comment:
This has the changes from #4594 but is difficult to review due to move. Probably best to review on the standalone PR.